Feature/2267 standalone functions

martincerny opened the pull request:
#### Submisison Checklist

- [x] Run unit tests: `./runTests.py src/test/unit`
- [x] Run cpplint: `make cpplint`
- [x] Declare copyright holder and open-source license: see below

#### Summary
Add support for standalone function compilation. See #2267. 

There are a lot of outstanding questions about whether this is the right way to implement it. I am copying those from #2267, as they seem to got lost there:

1) The decision to choose mode of compilation (functions only / model) by file extension and the choice of the ```.stanfuncs``` extension are both important and should be discussed by the core team (there could be a need for more compilation modes in the future, especially with regards to #2224).
    - My reasoning behind this decision was: 
      - The mode of compilation should be deducible from the file to be compiled alone and should not require the user to check some external configuration/build/... file. This is good for both reading code and editor support.
       - Deciding what to do with a file based on its extension is a common practice, in particular it is what gcc does.
2) I did some changes/refactoring to some very core files and to the test infrastructure to avoid duplicating code, but I felt very uncomfortable making the changes without supervision from someone who understands the project better.
3) Test functions files now reside in ```test/test-models/good-standalone-functions``` - is this a reasonable place in the tree? (an obvious alternative would be ```test/test-standalone-functions/good```)
    - Note that I cannot put ```.stanfuncs``` files simply in  ```test/test-models/good``` because it is then impossible to distinguish the resulting ```.hpp``` files from each other (and I don't think ```.stanfuncs``` should produce something like ```.hpp-funcs```)
4) I had trouble with the 80 char limit, I don't like some of the indents / line breaks it forced me to do, but I couldn't find a better way - could you please check that the code reads nice and is aligned with what others write? (e.g. the two-line ```for``` cycle in ```stanc_helper.hpp``` line 293).
5) When compiling standalone functions, it is definitely necessary to check whether other blocks are contained in the input file. I see two possible ways of doing this - what would you prefer?
    1) keep the default parser and check whether the members of ```stan::lang::program``` are all empty except ```function_decl_defs_``` (requires maintanence when new types of program blocks are introduced or when the ```program``` struct changes)
    2) create a specific parser variant that only recognizes the functions block (would require maintanance when the parser structure changes and a specific parser for other compilation modes would have to be created as well)
6) Should ```compile_functions()``` give a warning or error when other blocks are present?

#### Intended Effect
Add .stanfuncs as a possible source file extension that then results in compiling a file with instantiations of all templated functions.

#### How to Verify
.stanfuncs files are compiled as part of test build, also tests in ```src/test/unit/lang/parser/functions_standalone_instantiate_test.cpp``` and ```src/test/unit/lang/parser/functions_standalone_test.cpp```

#### Side Effects
Touched a lot of core files, please verify it is OK.

#### Documentation
Is rudimentary and pending approval of the core design.

#### Reviewer Suggestions

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

bob-carpenter commented:
Header tests failed.  Please let me know when you think this is ready to review and I can look it over.  I can do that before the header tests get fixed if it's otherwise done.  And thanks for submitting a pull request.

martincerny commented:
My bad, forgot I could run header tests locally. Fixed (hopefully).

bob-carpenter commented:
Thanks.  I'll get to reviewing this week.

mitzimorris commented:
I checked out this branch from Martin's repo.  As a test I took the file `src/test/test-models/good-standalone-functions/basic.stanfuncs` and copied it to `src/test/test-models/good/basic.stan`.  Also reformatted it for consistent indentation and placement of open curly brackets, so contents of both files is:
```
functions {
  real my_log1p_exp(real x) {
    return log1p_exp(x);
  }

  vector my_vector_mul_by_5(vector x) {
    vector[num_elements(x)] result = x * 5.0;
    return result;
  }
}
```

I compiled both files using Stan's makefile, which invokes stanc on the files `basic.stanfuncs` and `basic.stan:` 
```
make src/test/test-models/good-standalone-functions/basic.hpp-test
make src/test/test-models/good/basic.hpp-test
```
To compare these files, I edited out the model definition and the model typedef statement which is the last line of the transpiled stan program, then ran `diff`:

```
diff -U 3 src/test/test-models/good/basic.hpp src/test/test-models/good-standalone-functions/basic.hpp
--- src/test/test-models/good/basic.hpp	2017-05-28 11:50:02.000000000 -0400
+++ src/test/test-models/good-standalone-functions/basic.hpp	2017-05-28 11:51:13.000000000 -0400
@@ -1,24 +1,19 @@
 // Code generated by Stan version 2.15.0
 
-#include <stan/model/model_header.hpp>
-
-namespace basic_model_namespace {
+#include <stan/model/standalone_functions_header.hpp>
 
+namespace basic_functions { 
 using std::istream;
 using std::string;
 using std::stringstream;
 using std::vector;
-using stan::io::dump;
 using stan::math::lgamma;
-using stan::model::prob_grad;
 using namespace stan::math;
 
 typedef Eigen::Matrix<double,Eigen::Dynamic,1> vector_d;
 typedef Eigen::Matrix<double,1,Eigen::Dynamic> row_vector_d;
 typedef Eigen::Matrix<double,Eigen::Dynamic,Eigen::Dynamic> matrix_d;
 
-static int current_statement_begin__;
-
 template <typename T0__>
 typename boost::math::tools::promote_args<T0__>::type
 my_log1p_exp(const T0__& x, std::ostream* pstream__) {
@@ -91,6 +86,14 @@
     }
 };
 
+double
+my_log1p_exp(const double& x, std::ostream* pstream__){
+	return my_log1p_exp<double>(x, pstream__);
+}
 
+Eigen::Matrix<double, Eigen::Dynamic,1>
+my_vector_mul_by_5(const Eigen::Matrix<double, Eigen::Dynamic,1>& x, std::ostream* pstream__){
+	return my_vector_mul_by_5<double>(x, pstream__);
 }
 
+ } 
```

Two things:

1. why use a different header?
compiled stan program header file include:  
```
#include <stan/model/model_header.hpp>
```
compiled stan functions-only block header file include:
```
#include <stan/model/standalone_functions_header.hpp>
```

The set of includes in  `standalone_functions_header.hpp` is a subset of the files in `model_header.hpp`.  It would be better to just use the `model_header.hpp` file - first off, it's not clear to me that the standalone functions header has all necessary includes, secondly, maintaining two sets of includes is more work.

I reviewed the changes to the code in the `src/stan/lang/` and it seems OK.

However, we need clarification from @bgoodri as to what the comments need to be generated so that the transpiled hpp files can be included directly from R.

mitzimorris commented:
Please add test models for rng and lp functions.
Suggested rng function test:
```
functions {
  real foo_rng(real mu, real sigma) {
    return normal_rng(mu, sigma);
  }
}
```

bob-carpenter commented:
Did we also want a test that both ode_integrators work?

bgoodri commented:
At the top, it needs `// [[Rcpp::depends(rstan)]]`. Then, it needs `// [[Rcpp::export]]` without a semicolon before each function (but not the corresponding functor), as in
```
// [[Rcpp::export]]
inline double
my_function(...)
```
I would also suggest to add a default argument of the null pointer to each `std::ostream* pstream__` argument because that cannot be passed from R. For `_lpdf` and `_lpmf` functions, the trailing `lp_accum__` should be defaulted to zero. The functions that end in `_rng` are difficult. I currently replace the `base_rng__` argument with an `int seed` and then at the first line of the function body, I have
```
static boost::ecuyer1988 base_rng__(seed);
```

sakrejda commented:
You can pass both the RNG and stream of you have helper functions for
producing them.

On Sun, May 28, 2017, 9:00 PM bgoodri <notifications@github.com> wrote:

> At the top, it needs // [[Rcpp::depends(rstan)]]. Then, it needs //
> [[Rcpp::export]] without a semicolon before each function (but not the
> corresponding functor), as in
>
> // [[Rcpp::export]]
> inline double
> my_function(...)
>
> I would also suggest to add a default argument of the null pointer to each std::ostream*
> pstream__ argument because that cannot be passed from R. For _lpdf and
> _lpmf functions, the trailing lp_accum__ should be defaulted to zero. The
> functions that end in _rng are difficult. I currently replace the
> base_rng__ argument with an int seed and then at the first line of the
> function body, I have
>
> static boost::ecuyer1988 base_rng__(seed);
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/stan-dev/stan/pull/2310#issuecomment-304552033>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAfA6QmN31TLE45r530NzTSCRaN6bWQFks5r-hiPgaJpZM4Nfc1M>
> .
>

martincerny commented:
Thanks for the feedback, it is appreciated. However, all of the comments so far deal with (in my view) minor technical stuff. Does that mean that you are OK with the high-level structure and the main design decisions behind the way I implemented this feature? Especially the points 1-3 in the pull request description. Any opinion on 5&6 would also be welcome.

@mitzimorris The reduced set of headers was explicitly requested (https://github.com/stan-dev/stan/issues/2267#issuecomment-292600155) and should lead to shorter compilation times - the model includes are IMHO a huge beast. I currently don't have a strong opinion whether this is worth the extra maintanence cost, just wanted to point out the reasoning behind this choice.

martincerny commented:
@mitzimorris the test for ```_rng```, ```_lp``` and ```_lpdf``` functions was already there, in ```src/test/test-models/good-standalone-functions/special_functions.stanfuncs``` and it is tested that it both compiles and executes (in ```src/test/unit/lang/parser/functions_standalone_test.cpp``` and ```src/test/unit/lang/parser/functions_standalone_instantiate_test.cpp```. Is it sufficient or do you suggest extending it?

@bgoodri I added the RCpp comments and the null default to pstream (the default was added to all functions - even in normal Stan - is that a problem?), however, I cannot reasonably give default the ```lp_accum__``` argument as it is a ```double&```. I have also added a default ```__create_rng(int seed)``` function that is also exported to RCpp (currently for standalone functions only).

martincerny commented:
OK, so adding default for the ```pstream__``` argument turned out to be harder than expected: if a forward declaration is present, the default needs to be put there and only there. But AFAIK, I cannot check in the generator whether the function I am currently generating has a forward declaration somewhere above. 

So I added the default only to instantiations, but that means that it will not be present for functions which do not need instantiations (e.g. int-only arguments). Maybe it will be better to create a helper function to get a stream as well?

sakrejda commented:
BTW: this is more or less what the helper function in R would look like, similar for the stream (you could even pass `Rcpp::rcout` from the helper so the output would appear correctly.

```
#include <Rcpp.h>

// [[Rcpp::export]]
Rcpp::XPtr<boost::ecuyer1988> get_rng(const int& seed=0) {
  boost::ecuyer1988 base_rng__(seed);
  Rcpp::XPtr<boost::ecuyer1988> ptr(base_rng__);
  return ptr;
}
```

mitzimorris commented on src/stan/lang/generator/generate_rng_rcpp_helper.hpp:
```diff
@@ -0,0 +1,26 @@
+      o << "// [[Rcpp::export]]" << EOL;
+      o << "boost::ecuyer1988 __create_rng(int seed) {" << EOL;
+      o << "\treturn(boost::ecuyer1988(seed));" << EOL;
+      o << "}" << EOL << EOL;
```
what is different between this and suggestion:
```
  Rcpp::XPtr<boost::ecuyer1988> ptr(base_rng__);
  return ptr;
```

bob-carpenter commented:
No, we can't do what @sakrejda is suggesting in https://github.com/stan-dev/stan/pull/2310#issuecomment-304935311 because we can't introduce dependencies on R or Rcpp (or any other GPL-ed code) in Stan itself.

mitzimorris commented on src/stan/lang/generator/generate_function_instantiation_body.hpp:
```diff
@@ -0,0 +1,68 @@
+                                    std::ostream& o) {
+      o << "{" << EOL;
+      o << "\t";
+      if (!fun.return_type_.is_void()) {
```
replace all `\t` with 2 spaces - generated code should contain spaces not tabs - change this everywhere.

sakrejda commented:
Bob, I'm not suggesting putting Rcpp stuff into Stan code. The point is we
can leave the RNG argument in these generated functions because it's
trivial for rstan to provide a function that fills it in!

On Tue, Jun 13, 2017, 4:31 PM Mitzi Morris <notifications@github.com> wrote:

> *@mitzimorris* requested changes on this pull request.
> ------------------------------
>
> In src/stan/lang/generator/generate_function_instantiation_body.hpp
> <https://github.com/stan-dev/stan/pull/2310#discussion_r121788601>:
>
> > +     * @param[in] fun function declaration
> +     * @param[in] is_rng true if function is an RNG
> +     * @param[in] is_lp true if function accesses log density
> +     * accumulator
> +     * @param[in] is_log true if function is log probability function
> +     * @param[in] rng_class class of the RNG being used (required by xxx_rng
> +     * functions)
> +     * @param[in,out] o stream for generating
> +     */
> +    void generate_function_instantiation_body(const function_decl_def& fun,
> +                                    bool is_rng, bool is_lp, bool is_log,
> +                                    const std::string& rng_class,
> +                                    std::ostream& o) {
> +      o << "{" << EOL;
> +      o << "\t";
> +      if (!fun.return_type_.is_void()) {
>
> replace all \t with 2 spaces - generated code should contain spaces not
> tabs - change this everywhere.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/stan-dev/stan/pull/2310#pullrequestreview-43846067>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAfA6bcwKtt8KIZRAMIeUlA2FDl4vHkeks5sDvG3gaJpZM4Nfc1M>
> .
>

bob-carpenter commented:
@sakrejda  Great --- as long as the helper R function stays in RStan, that's fine.

bob-carpenter commented:
I'd go and make these changes myself, but I don't know how to work on remote branches.  @seantalts --- you said you'd show me how to do this, but I'm going to be out of town for a week.  Would it be easy for me to somehow change all the `\t` to two spaces?  I think that's the last thing waiting to do before this gets merged.

martincerny commented:
I've replaced the tabs with spaces. Should be ready to go.
